### ISSUE TITLE
mc mirror takes long time when target bucket contains a lot of files

### ISSUE THREAD

**Issue by archsh** [minio/mc#4873]
## Expected behavior
Using `mc mirror` to mirror local folder to a target bucket, it should be exit soon if local folder is empty or few files.

## Actual behavior
It took a long time to start transfer files if the target bucket contains a lot of objects.

## Steps to reproduce the behavior

mc mirror --json ./localfolder     target/bucket/path

## mc --version
- (paste output of `mc --version`)
mc version RELEASE.2024-01-05T05-04-32Z (commit-id=59eca9fea8984adec1e8e7a1c95d0ea23107ceff)

## System information

CentOS release 6.9 (Final) x64

**Comment by zveinn** [comment]
Yes, mc mirror does a full source to target comparison when it's ran. This is due to the --remove flag. However, it might be possible to speed up the mirror given that the --remove flag is not present. I will look into this in the coming weeks, no promises on delivery date though since my schedule is rather full.

**Comment by zveinn** [comment]
I will be looking into this next week, if there is any additional information you want to share, then now is the time :)

**Comment by karsapeng** [comment]
helloHello, I wrote a script to use mc mirror --remove --overwrite --limit-download $LIMIT_DOWNLOAD --limit-upload $LIMIT_UPLOAD minio-master minio-slave to synchronize the data of the master node. Even if the data of the two minio nodes are consistent, if the master node is shut down and restarted due to a fault, I will change the direction of the synchronization and transfer the data to the slave node. Then the master will run mc mirror --remove --overwrite --limit-download $LIMIT_DOWNLOAD --limit-upload $LIMIT_UPLOAD minio-slave minio-master for synchronization. However, even if the data of the two minio nodes is consistent, minio-master will perform a full synchronization, which will take up a lot of io and cause the server to crash. My mc version is mc:RELEASE.2024-08-26T10-49-58Z

**Comment by karsapeng** [comment]
Hello, has this problem been solved?

**Comment by zveinn** [comment]
Hey, I have not had time to look into this yet. Just as I was about to start, something else came up.  I will see if someone else can take it off my hands.

**Comment by karsapeng** [comment]
hello，Is there any progress on this issue？I'm a little troubled with this mirror function  :)

**Comment by fherenius** [comment]
One thing to add, I’m using `mc mirror` to move a cluster. The target cluster already has several million objects in it.

`mc mirror` will do a full synchronization of the target, even for buckets that don’t exist on the source. (Visible when running with `—debug`). It would make it a lot faster if it would filter buckets on name first, instead of content.

Or maybe, `—exclude-bucket` should also count for the target synchronization.

**Comment by zveinn** [comment]
@fherenius @karsapeng @PrisedRabbit @archsh 
Sorry how long this took, minio is growing fast and we've been super busy during the last few months. 

https://github.com/minio/mc/pull/5116

This allows mc to exit early when the local bucket has synced when `--remove` is not used.

**Comment by zveinn** [comment]
The PR should be ready now, please test it when you get the time.

**Comment by Laisky** [comment]
I need to synchronize near millions files in a bucket, but `mc mirror` has completely stopped responding and has even crashed the Minio server. I think it is absurd to compare the entire file lists of the source and target. Instead, we should use a BFS method to compare small portions of the file list layer by layer to effectively manage such a large number of files.

cc @zveinn

**Comment by zveinn** [comment]
It won't matter if we do the comparison in batches, since we will have to cover the entire space anyways. We don't check every object on the target, unless you have --remove toggled. Then we do need to check them all. ( I made a fix for this recently ) https://github.com/minio/mc/pull/5116/files

If mc is crashing the minio server then it's an I/O problem, maybe you can use the --limit-upload/download to ensure you don't overload your drives ?

**Comment by Laisky** [comment]
In certain situations, a MinIO bucket may contain so many files that the server cannot handle the `mc ls -r` operation, which leads to `mc mirror` freezing completely. Even though not every object is being fetched, the `ls -r` operation can overwhelm the server. Introducing a layer-by-layer BFS search can enable the execution of `mc mirror` on a vast number of files in low-performance servers.

I have submitted a pull request that introduces the `--bfs` option for `mc mirror`, which implements this functionality.

cc @zveinn
